File: aungnanda_oo/q7_KAnagrams.py
# Question 7: KAnagrams

# Two strings are considered to be “k-anagrams” if they can be made into anagrams by changing at most k characters in one of the strings. Given two strings and an integer k, determine if they are k-anagrams.

# Examples:

# Input Strings: "apple", "peach"
# Input k: 1
# Output: False

# Input Strings: "apple", "peach"
# Input k: 2
# Output: True

# Input Strings: "cat", "dog"
# Input k: 3
# Output: True

# Input Strings: "debit curd", "bad credit"
# Input k: 1
# Output: True

# Input Strings: "baseball", "basketball"
# Input k: 2
# Output: False

from collections import Counter
def KAnagrams(s1, s2, k) -> bool:

    if len(s1) != len(s2):
        return False
    
    c1 = Counter(s1)
    c2 = Counter(s2)


    changes = 0
    for c in c1:
        if c1[c] > c2.get(c, 0):
            changes += c1[c] - c2.get(c, 0)

    return changes <= k

# Time Complexity: O(n)
# Space Complexity: O(n)

test_cases = [("apple", "peach",1), ("apple", "peach", 2), ("cat", "dog",3), ("debit curd", "bad credit", 1), ( "baseball", "basketball", 2) ]

for test_case in test_cases:
    s1, s2, k = test_case
    print(KAnagrams(s1, s2, k)) 

# Spent a total of 40 mins on this question

File: homework1/steven_law/q7_KAnagrams.cpp
/*Time Complexity:O(n)
Space Complexity:O(1) 26 letters in alpha 
Technique used: Hashing
Time Taken: 10mins 46 secs
*/

#include<iostream>
#include<vector>
#include<unordered_map>
#include <string>
#include <unordered_set>

using namespace std;

bool kAnagrams(string s1, string s2, int k){
    if (s1.length() != s2.length()) return false;

    int changesNeeded = 0;
    unordered_map<char, int> count;
    for(char letter: s1){
        count[letter]++;
    }
    for(char letter: s2){
        count[letter]--;
    }
    for(auto const& [letter, counts] : count){
        if(counts>0){
        changesNeeded+= counts;
        }
    }
    if(changesNeeded <= k) return true;
    return false;
}

/* psudocode/thoughts/logic
goal: Determine if two strings are k-anagrams
restraints: Must be same length strings.
strategy: My strategy going into this is just to sort each string and 
compare in line with one another, rack up a count for the different 
characters or unique characters then compare that to k, if less or = to 
then it works, if not then it does not. this is slow though. O(n log n)
to sort. we can try hashing. we can use a similar approach instead of 
sorting we can keep a count of the frequency of each character in the first
string and comapre it to the second string and calculate the difference
and compare that to k.

*/

int main(){
    string s1 = "apple", s2 = "peach";
    int k = 1;
    cout << "Expected: False" << endl;
    cout << "Output:   " << (kAnagrams(s1, s2, k) ? "True" : "False") << endl << endl;

    s1 = "apple"; s2 = "peach"; k = 2;
    cout << "Expected: True" << endl;
    cout << "Output:   " << (kAnagrams(s1, s2, k) ? "True" : "False") << endl << endl;

    s1 = "cat"; s2 = "dog"; k = 3;
    cout << "Expected: True" << endl;
    cout << "Output:   " << (kAnagrams(s1, s2, k) ? "True" : "False") << endl << endl;

    s1 = "debit curd"; s2 = "bad credit"; k = 1;
    cout << "Expected: True" << endl;
    cout << "Output:   " << (kAnagrams(s1, s2, k) ? "True" : "False") << endl << endl;

    s1 = "baseball"; s2 = "basketball"; k = 2;
    cout << "Expected: False" << endl;
    cout << "Output:   " << (kAnagrams(s1, s2, k) ? "True" : "False") << endl;

    return 0;
}
